sdks/python/apache_beam/runners/interactive/interactive_environment.py

```diff
@@ -163,7 +165,8 @@ def __init__(self):
     # the gRPC server serves.
     self._test_stream_service_controllers = {}
     self._cached_source_signature = {}
-    self._tracked_user_pipelines = set()
```

Comment:
LGTM, only one question: 
Does the `UserPipelineTracker` clean up the user pipeline and their derived pipelines when a user pipeline is out of scope (e.g., deleted or garbage collected). 

If not, is there any side effect when the user uses an outdated pipeline ref or a new pipeline ref (from re-executions) that results in the same `__hash__` or `__eq__`/`in` to be `True`? Will that give back a wrong user_pipeline when the tracker thinks the pipeline is tracked while it's not?

Reply:
Ack, added eviction capabilities. The `__hash__` value is based on the id of the object. If the user still has a ref to the pipeline then the hash will still be unique. Your latter paragraph probably wouldn't occur in that case.